### PR TITLE
bugs fixed for lights, material, and geometry

### PR DESCRIPTION
**Pull request by karenpeng**
-bugs fixed:
 - color parameters parsing in light
 - refine sphere and torus normals

### CODE REVIEW

**Comment by indefinit** [comment]
Hey @karenpeng thanks for submitting this.  Going to make some comments inline.  Generally speaking, I'm wondering if we can avoid splitting our modelview matrix into two matrices as it adds a bit more overhead on both the CPU and shader sides.  Also, perhaps we can try to emulate the existing Processing api here:  https://processing.org/reference/beginCamera_.html

Lastly, if you could provide more info in your commit message so I can better understand what the bugs are being fixed that would be very helpful!  Perhaps a link to a test of some sort?

File: src/webgl/light.js [hunk 1]
```diff
@@ -150,16 +150,17 @@ p5.prototype.directionalLight = function(v1, v2, v3, a, x, y, z) {
 
   var _x, _y, _z;
 
-  if(typeof arguments[arguments.length-1] === 'number'){
-    _x = arguments[arguments.length-3];
-    _y = arguments[arguments.length-2];
-    _z = arguments[arguments.length-1];
+  var argus = Array.prototype.slice.call(arguments);
```
Comment (indefinit) [inline src/webgl/light.js]: check out primitives.js for naming convention regarding arguments to Array.  variable should be var args instead of argus

File: src/webgl/p5.RendererGL.js [hunk 2]
```diff
@@ -44,7 +44,8 @@ p5.RendererGL = function(elt, pInst, isMainCanvas) {
    * model view, projection, & normal
    * matrices
    */
-  this.uMVMatrix = new p5.Matrix();
+  this.uVMatrix = new p5.Matrix();
```
Comment (indefinit) [inline src/webgl/p5.RendererGL.js]: Explained more in the general comment, but I'd prefer not to have separate view matrices and model matrices if possible.

File: src/webgl/p5.RendererGL.js [hunk 3]
```diff
@@ -478,26 +488,36 @@ p5.RendererGL.prototype.rotateZ = function(rad) {
...
 
 p5.RendererGL.prototype.resetMatrix = function() {
-  this.uMVMatrix = p5.Matrix.identity();
+  this.uMMatrix = p5.Matrix.identity();
+  this.uVMatrix = p5.Matrix.identity();
   this.translate(0, 0, -800);
   return this;
 };
 
+p5.RendererGL.prototype.cameraTranslate = function(x, y, z) {
```
Comment (indefinit) [inline src/webgl/p5.RendererGL.js]: perhaps we can adopt an approach similar to: https://processing.org/reference/beginCamera_.html where translations and rotations are simply applied to a matrix in our MVMatrix stack?

File: src/webgl/material.js [hunk 4]
```diff
@@ -178,7 +178,7 @@ p5.prototype.ambientMaterial = function(v1, v2, v3, a) {
   gl.useProgram(shaderProgram);
   shaderProgram.uMaterialColor = gl.getUniformLocation(
     shaderProgram, 'uMaterialColor' );
-  var colors = this._renderer._applyColorBlend(v1,v2,v3,a);
+  var colors = this._renderer._applyColorBlend.apply(this._renderer, arguments);
```
Comment (indefinit) [inline src/webgl/material.js]: I'm not sure if I see where the bug is here.  apply is called on the current renderer context within _applyColorBlend.  If we want to call apply outside of the _applyColorBlend function, we should rename _applyColorBlend to _blendColor and tweak that function so as not to apply separate context inside.

Comment (karenpeng) [inline src/webgl/material.js]: @indefinit, for example `specularMaterial(250)`: `v1, v2, v3, a` is parsed in, the arguments in `_applyColorBlend` will become `[250, undefined, undefined, undefined]` instead of `[250]`. It will be translated to be red, instead of gray.
That's why we are seeing a red color in here http://p5js.org/examples/examples/3D_Multiple_Lights.php.

File: src/webgl/p5.RendererGL.js (hunk 3 above)
Comment (karenpeng) [inline src/webgl/p5.RendererGL.js]: I'll explain it in more details later. I think it could be a great conversation.

File: src/webgl/material.js (hunk 4 above)
Comment (indefinit) [inline src/webgl/material.js]: @karenpeng Sorry for the delay getting back to you on this.  Yes, good point.  Could you pull this bug fix into a separate PR and please submit?  I think the camera changes are a larger conversation, but you're 100% right about the _applyColorBlend bug and I'd like to get this patched ASAP.  Thanks!

**Comment by karenpeng** [comment]
@indefinit, 
Here's the modified diff with just the bugs fixed, the separation of model and view matrix is removed.

**Comment by indefinit** [comment]
Hi @karenpeng thanks for making the changes.  This is almost there!  Only comment now is please replace calls to Array.prototype.slice with the original:
```
var args = new Array(arguments.length);
for (var i = 0; i < args.length; ++i) {
   args[i] = arguments[i];
}
```
The latter is more performant than Array.prototype.slice.

**Comment by karenpeng** [comment]
@indefinit, [mdn](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/arguments) suggests this:
Using slice on arguments prevents optimizations in some JavaScript engines (V8 for example - more information). If you care for them, try constructing a new array by iterating through the arguments object instead. An alternative would be to use the despised Array constructor as a function:
```js
var args = (arguments.length === 1 ? [arguments[0]] : Array.apply(null, arguments));
```
How about we go for that? It's more clean and readable.

**Comment by indefinit** [comment]
I think we should follow the general conventions set in p5 already.  Here's a PR from last year discussing arguments optimizations: https://github.com/processing/p5.js/pull/938

Ultimately we ended up with the existing args approach with for loop instead of array.prototype.slice

**Comment by karenpeng** [comment]
@indefinit , ahh thanks for pointing out the conversation. Fixed it.

**Comment by lmccart** [comment]
thanks @karenpeng! @indefinit feel free to merge this whenever you feel it's ready.
